ledger-service/http-json/src/main/scala/com/digitalasset/http/Endpoints.scala

```diff
@@ -120,38 +122,65 @@ class Endpoints(
                 )
               )
           }
-        } yield response
+        } yield transformedResponse
       }
     }
 
   private def logJsonRequestAndResult(implicit
       lc: LoggingContextOf[InstanceUUID with RequestID]
   ): Directive0 = {
-    def logWithHttpMessageBodyIfAvailable(
-        httpMessage: HttpMessage,
+    def logWithHttpMessageBodyIfAvailable[A <: HttpMessage](
+        httpMessage: A,
         msg: String,
-        bodyKind: String,
-    ): Future[Unit] =
+        kind: String,
+    ): HttpMessage =
       if (
         httpMessage
           .header[`Content-Type`]
           .map(_.contentType)
           .contains(ContentTypes.`application/json`)
-      )
-        httpMessage
-          .entity()
-          .toStrict(maxTimeToCollectRequest)
-          .map(it =>
+      ) {
+        httpMessage.entity.contentLengthOption match {
+          // Limit logging of bodies to content with size of less than 10 KiB
+          case Some(length) if length < 1024 * 10 =>
+            import akka.stream.scaladsl._
+            logger.debug("Calling transformEntityDataBytes")
+            httpMessage
+              .transformEntityDataBytes(
+                Flow.fromFunction { it =>
+                  try {
+                    withEnrichedLoggingContext(
+                      LoggingContextOf.label[RequestEntity],
+                      s"${kind}_body" -> it.utf8String.parseJson,
+                    )
+                      .run { implicit lc => logger.info(msg) }
+                  } catch {
+                    case ex: Exception =>
+                      logger.error("Failed to log message body: ", ex)
+                  }
+                  it
+                }
+              )
+          case other =>
+            val ctxContent = other
+              .map(length => s"OMITTED BECAUSE BODY SIZE OF $length IS TOO BIG FOR LOGGING")
+              .getOrElse {
+                if (httpMessage.entity.isChunked())
+                  "OMITTED BECAUSE REQUEST BODY IS CHUNKED & OVERALL SIZE IS UNKNOWN"
+                else
+                  "OMITTED BECAUSE REQUEST BODY SIZE IS UNKNOWN"
+              }
             withEnrichedLoggingContext(
               LoggingContextOf.label[RequestEntity],
-              s"${bodyKind}_body" -> it.data.utf8String.parseJson,
-            )
-              .run(implicit lc => logger.info(msg))
-          )
-          .recover { case ex =>
-            logger.error("Failed to extract body for logging", ex)
-          }
-      else Future.successful(logger.info(msg))
+              s"${kind}_body" -> ctxContent,
+            ).run { implicit lc => logger.info(msg) }
+            httpMessage
+        }
+      } else {
+        logger.info(msg)
+        httpMessage
+      }
```

Comment:
There's a lot of nesting and apparently some repetition within these method. Would it be possible to factor out those parts that seem to do similar things

1.
```
                    withEnrichedLoggingContext(
                      LoggingContextOf.label[RequestEntity],
                      s"${kind}_body" -> it.utf8String.parseJson,
                    )
                      .run { implicit lc => logger.info(msg) }
```
2.
```
              s"${kind}_body" -> ctxContent,
            ).run { implicit lc => logger.info(msg) }
            httpMessage
```
3.
```
        logger.info(msg)
        httpMessage
```

and possibly reduce the nesting by extracting methods and flattening the nested `other.map` into part of the `match` expression?

Reply:
I agree that this can be improved. Will reduce duplication here as possible.